MPI_*_get_info/set_info()

markalle opened the pull request:
This is built off of Dave Solt's earlier pull request #1308.

Roughly what Dave did is:
- change ```ompi_{communicator,win,file}_t``` to have ```opal_infosubscriber_t``` as their ```.super``` class
- change ```ompi_info_t``` to have ```opal_info_t``` as its ```.super``` class
- transition a lot of the ```ompi_info_*``` calls to ```opal_info_*```

I fixed memory leaks, compiler warnings, and changed the semantics to support the current interpretation of the MPI standard.

The main functions are
    ```opal_infosubscribe_infosubscribe()``` : register a callback fn for a particular key
    ```opal_infosubscribe_change_info()``` : used at ```MPI_*_set_info()``` time
    ```opal_info_dup_mpistandard()``` : extracts the k/v as defined for ```MPI_*_get_info()```

jjhursey commented:
Ref PR #1308

ibm-ompi commented:
Build Failed with PGI compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/780c3f2fcc4260c39e3f000f2d346cc3

jjhursey commented:
bot:ibm:pgi:retest

ibm-ompi commented:
Build Failed with PGI compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/0f51d7bc06a4ab54bd6a7adeed9f5ded

jjhursey commented:
I originally thought that the PGI CI failure was a local system issue. However, it seems legit, as best I can tell.

```
  CC       src/io_romio314_module.lo
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-W-0093-Type cast required for this conversion of constant (src/io_romio314_module.c: 108)
PGC-S-0073-Too many initializers for mca_io_romio314_module (src/io_romio314_module.c: 108)
PGC/power Linux 16.10-0: compilation completed with severe errors
```

jjhursey commented:
Yeah it looks like romio needs to be updated in this PR too. I'm not sure why the PGI compiler is the sensitive one here, but it's a good catch.

The CI configures with:
```shell
./configure --without-x CC=pgcc CXX=pgc++ FC=pgfortran
```

@markalle If you need access to the PGI compiler just ping me and I'll let you know where they are on the ppc64le dev machines.

jsquyres commented on AUTHORS:
```diff
@@ -81,6 +80,8 @@ Dave Goodell, Cisco
 David Daniel, Los Alamos National Laboratory
   ddd@lanl.gov
+David Solt, IBM
+  dsolt@us.ibm.com
```
Note that `AUTHORS` is generated automatically these days.

markalle replied:
I reverted it anyway, so dave's checkin now adds a line, and my checkin removes it so the two commits together should be a no-op on AUTHORS

jsquyres commented on ompi/communicator/comm.c:
```diff
@@ -22,6 +22,7 @@
  *                         and Technology (RIST). All rights reserved.
  * Copyright (c) 2014-2015 Intel, Inc. All rights reserved.
  * Copyright (c) 2015      Mellanox Technologies. All rights reserved.
+ * Copyright (c) 2016 IBM Corp.  All rights reserved.
```
Shouldn't these copyright updates be 2017 now?

jsquyres commented on ompi/mca/io/romio314/src/io_romio314_component.c:
```diff
@@ -234,15 +235,29 @@ static int delete_query(const char *filename, struct ompi_info_t *info,
+// I don't think there's any advantage to most of the code being switched
+// to opal_info_t, which is causing the unnecessary extra complexity here:
+    ompi_info_t *ompi_info;
+    ompi_info = OBJ_NEW(ompi_info_t);
```
Need to check for NULL here.

markalle replied:
Thanks, I made a change for checking that NULL

jsquyres commented on ompi/mca/io/romio314/src/io_romio314_component.c:
```diff
@@ -234,15 +235,29 @@ static int delete_query(const char *filename, struct ompi_info_t *info,
+// I'd be open to reverting most of it back to ompi_info_t and using
+// opal_info_* calls on the opal_info_t that's inside the ompi_info_t.
+// I don't think there's any advantage to most of the code being switched
+// to opal_info_t, which is causing the unnecessary extra complexity here:
```
Seems like this comment should be addressed...? (a similar comment is repeated several times)

markalle replied:
I guess a decision needs made for what is the predominant design. Dave has ompi_info_t as the larger class that contains an opal_info_t plus a tiny bit extra (like a fortran int id). All the top level MPI calls that use info come in with an ompi_info_t*.

Dave's overall design is to migrate toward the code from that point down being largely opal_info_t. I don't really disagree, although it does end up with an awful lot of
  ```some_function(mpiinfo)```  becoming  ```some_function(&mpiinfo->super)``` // to get the opal_info_t
and equivalently, if you just want to pass MPI_INFO_NULL to some routine, but it wants an opal_info_t*, you have to give it &MPI_INFO_NULL->super.

And there are a tiny number of places where the lower level code calls back into MPI, and thus needs the larger ompi_info_t when all it has is an opal_info_t, thus having the extra code you quoted that's making another copy of the info that isn't logically necessary. That bothers me, but it's only 3 places... so I guess I'm leaning toward reverting my comment, and sticking with the design of having the lower level code predominantly using opal_info_t.

ibm-ompi commented:
Build Failed with PGI compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/56af0cc8bee2c10cc19913afd2737780

gpaulsen commented:
PGI compiler issues is fixes, Jeff's review comments have been updated.

We think this is ready to merge to master.  Can we merge it in, so that @hjelmn can begin using this?  If the MPI Forum changes their minds on open questions about this, we can address those in a separate PR.

jjhursey commented:
bot:ibm:retest

gpaulsen commented:
@jsquyres can you please re-review?

jjhursey commented:
@markalle It looks like this branch needs to be rebased again. Can you do that so @jsquyres can review again?

ibm-ompi commented:
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/7a1f378bca3c4e57a1bf40f857338b69

gpaulsen commented:
We'd like to come to a conclusion on this discussion and close this out.
It needs to be rebased (again).  If MPI_Forum changes it's current thinking, it should be easy to change.

jsquyres commented:
bot:lanl:retest

jsquyres commented:
bot:ibm:retest

ibm-ompi commented:
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/b318cdde19cffc6796c960dce0b5196c

jsquyres commented:
@gpaulsen @markalle IBM PGI compile testing is failing.  Can you guys have a look?

gpaulsen commented:
Absolutely.  The current theory is that after a rebase it might get resolved.  But time will tell.

jsquyres commented:
@gpaulsen Ok.  Then let's do the rebase before the next review (otherwise, it'll just have to be reviewed again, right?).

ibm-ompi commented:
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/220e7dd7add3485006056dc9aeb9545c

markalle commented:
I don't guess I've documented much about how this code is structured yet.

The order of events when an object (comm/file/win) is created should be

1. if there's an incoming info, it is duped onto the new object

for example, ompi/communicator/comm.c
```
+        // Copy info if there is one.
+        newcomp->super.s_info = OBJ_NEW(opal_info_t);
+        if (info) {
+            opal_info_dup(info, &(newcomp->super.s_info));
+        }
```

2. ```opal_infosubscribe_subscribe()``` happens on each key of interest

for example in ompi_osc_rdma_component_select()
```+    opal_infosubscribe_subscribe(&win->super, "no_locks", "false", ompi_osc_rdma_set_no_lock_info);```

The callback specified in the ```opal_infosubscribe_subscribe()``` is activated in two cases:

1. when it is being registered in ```opal_infosubscribe_subscribe()```
2. in ```MPI_*_set_info()```

The value inputted into the callback is saved at ```__IN_<key>/<val>```, so this ends up tracking either the initial default system-provided value or the most recent manual set_info value.

The ```opal_info_dup_mpistandard()``` call decides what to output when the user does a get_info.  For a given ```<key>```, if there's an ```__IN_<key>``` and a ```<key>``` the value associated with ```__IN_<key>``` is used. That results in the original user-provided value being used or the system default if no user-value was given. In both cases if callback() modified the value the modification isn't user-visible. The only visibility into callback() comes from rejecting a key (when the callback() returns NULL). In that case there would be an ```__IN_<key>``` but no ```<key>``` and in that case we would return nothing for that key and the user would know it was rejected.

markalle commented:
Hmm, I just rebased and the PGI build is still reporting
```PGC-S-0073-Too many initializers for ompi_osc_pt2pt_module_template (osc_pt2pt_component.c: 111)```
where I don't believe that section is modified by these commits.

gpaulsen commented:
@jsquyres, Mark has resolved the PGI issues, and rebased.  Can you please re-review.

gpaulsen commented:
@jsquyres Can you please take another look?

jsquyres commented:
Also, if the 2nd commit on this PR is just fixes for the first commit, might as well squash them down into a single commit.

markalle commented:
Okay, I removed the unused variables.

I don't think the two commons are really additions:
```
    info.o: 0000000000000100 C ompi_mpi_info_env
    info.o: 0000000000000100 C ompi_mpi_info_null
```
they were previously
```    ompi_predefined_info_t ompi_mpi_info_env = {{{{0}}}};```
and now are just
```    ompi_predefined_info_t ompi_mpi_info_env;```
so maybe that changed the storage area for them.

And they're used from mpi.h as
```    #define MPI_INFO_ENV OMPI_PREDEFINED_GLOBAL(MPI_Info, ompi_mpi_info_env)```
for example, so they're meant to be globals in some form.

On the length I agree, I introduced that bug and I can't see why I made that change, I reverted it.

markalle commented:
Oh, for the squash, yeah I'm okay with that. It's just since Dave did the initial work it had been convenient for me to be able to distinguish what he did from what I changed.

ibm-ompi commented:
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/fe5ed39c3b59af6a683826aa06d86eda

rhc54 commented:
you can squash selectively - so you can squash yours down to one, and his down to another

ibm-ompi commented:
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ab258c38d7e8e3fdf7a9f13d9ac24c66

ibm-ompi commented:
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/0ec3785261b70f70f047bbf895c89c2b

markalle commented:
The last rebase brought in a need for interlib.c to #include "mpi.h" so it could see MPI_THREAD_SINGLE. I don't know what changed in the chain of includes that used to let it see that, but that's what made the previous CI test fail, so I added that change and re-pushed.

markalle commented:
I've been keeping it squashed to two commits: one from dave, one from me. I think @jsquyres was proposing squashing dave's and mine together.

Having them separated was useful for example when Jeff noted that test 30_get.c was failing and I could see whether that bug was my fault or Dave's (it was mine). But it's not critical, so I can squash to one if you want.

rhc54 commented:
nah - tell @jsquyres to get a life and leave it as two 😄

jsquyres commented on ompi/dpm/dpm.c:
```diff
@@ -18,6 +18,7 @@
  * Copyright (c) 2013-2016 Intel, Inc.  All rights reserved.
  * Copyright (c) 2014-2017 Research Organization for Information Science
  *                         and Technology (RIST). All rights reserved.
+ * Copyright (c) 2016-2017 IBM Corporation. All rights reserved.
```
Any particular reason for this copyright addition?

jsquyres commented on ompi/file/file.h:
```diff
@@ -15,6 +15,7 @@
  * Copyright (c) 2015      Research Organization for Information Science
  *                         and Technology (RIST). All rights reserved.
  * Copyright (c) 2016      University of Houston. All rights reserved.
+ * Copyright (c) 2016 IBM Corp.  All rights reserved.
```
If this is a new copyright, shouldn't it be 2017?

jsquyres commented on ompi/mca/fs/lustre/fs_lustre_file_delete.c:
```diff
@@ -10,6 +10,7 @@
  * Copyright (c) 2004-2005 The Regents of the University of California.
  *                         All rights reserved.
  * Copyright (c) 2008-2011 University of Houston. All rights reserved.
+ * Copyright (c) 2016 IBM Corp.  All rights reserved.
```
Here's another "new" 2016 copyright.

jsquyres commented on ompi/mca/fs/lustre/fs_lustre_file_open.c:
```diff
@@ -12,6 +12,7 @@
  * Copyright (c) 2008-2015 University of Houston. All rights reserved.
  * Copyright (c) 2015      Research Organization for Information Science
  *                         and Technology (RIST). All rights reserved.
+ * Copyright (c) 2016 IBM Corp.  All rights reserved.
```
..another "new" 2016 copyright.  I'll stop noting these.  😄

markalle commented:
@jsquyres
Did you see my earlier explanation of those commons?

They were previously
```ompi_predefined_info_t ompi_mpi_info_env = {{{{0}}}};```
and now are just
```ompi_predefined_info_t ompi_mpi_info_env;```
so maybe that changed the storage area for them.

And they're used from mpi.h as
#define MPI_INFO_ENV OMPI_PREDEFINED_GLOBAL(MPI_Info, ompi_mpi_info_env)
for example, so they're meant to be globals in some form.

markalle commented:
I'm guessing the copyright additions that say 2016 started back then, like you said. And the unmodified files with copyright additions had some change we later reverted. I guess removing all 2016 additions wouldn't be that hard.

rhc54 commented:
I don't know - I think that's getting into a little gray area. True, those changes weren't merged until 2017, but that isn't when the copyright applies. The copyright attests to when those changes were written, not when they became public. So I think @jsquyres is incorrect here and those copyrights should stay the way they are.

That's why your IP lawyers always insist you enter dates into your "notebooks" - the legal origination of the idea is the date in the notebook, not the date of publication.

jsquyres commented:
@markalle Yes, I saw your explanations.  But one way or another, we should not be introducing new common symbols.  😄   I guess I'm asking: shouldn't these new `info` symbols be able to be like other `mpi.h` symbols (e.g., `ompi_mpi_comm_world`)?

markalle commented:
I pushed again, with another pass through the copyrights, removing about a dozen files where the copyright was the only change, and changing a bunch of 2016 to 2017

rhc54 commented:
As I said, I honestly don't believe the changes from 2016 to 2017 are legally correct - but I honestly also don't think it's important as the copyright notices don't really mean anything in this context.

markalle commented:
Well darn, I didn't see Ralph's copyright comments until I had already made the change. But fortunately I still had a list of every file my previous script had run on, so for every file that used to contain 2016 I switched it back to 2016-2017.

And I finally removed the commons. I figured there must have been some reason why Dave had removed the {{{{0}}}} initializer that used to be there, but I can't fathom any reason not to initialize it so I'm guessing it was just the inconvenience of the syntax of figuring out how many braces it should have.

gpaulsen commented:
:bot:help

markalle commented:
:retest:

hjelmn commented:
:bot:retest:
